### PR TITLE
Adds FillContextEdges workload for mlmd_bench.

### PR DESCRIPTION
**Pull request by BrianSong**

### CODE REVIEW

File: ml_metadata/tools/mlmd_bench/proto/mlmd_bench.proto [hunk 1]
```diff
@@ -35,6 +35,12 @@ message UniformDistribution {
   optional int64 maximum = 2;
 }
 
+// A categorical distribution generated w.r.t. a Dirichlet prior with
+// a concentration parameter `dirichlet_alpha`.
+message CategoricalDistribution {
+  optional double dirichlet_alpha = 1;
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/proto/mlmd_bench.proto:45]: + comment, the rationale of using the special case of the dirichlet .

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+ private:
+  // Workload configurations specified by the users.
+  const FillContextEdgesConfig fill_context_edges_config_;
+  // Number of operations for the current workload.
+  const int64 num_operations_;
+  // String for indicating the name of current workload instance.
+  const std::string name_;
+  // Records all the Context X Artifact pairs seen in current setup.
+  std::unordered_map<int64, std::unordered_set<int64>>
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_context_edges_workload.h]: nit, unordered-set of pair?

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.h [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      context_id_to_artifact_ids_;
+  // Records all the Context X Execution pairs seen in current setup.
+  std::unordered_map<int64, std::unordered_set<int64>>
+      context_id_to_execution_ids_;
+};
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_context_edges_workload.h]: drop one?

File: ml_metadata/tools/mlmd_bench/benchmark.cc [hunk 4]
```diff
@@ -42,6 +43,12 @@ void CreateWorkload(const WorkloadConfig& workload_config,
                     workload_config.num_operations()));
       break;
     }
+    case WorkloadConfig::kFillContextEdgesConfig: {
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/benchmark.cc:46]: add a test

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 5]
```diff
@@ -265,4 +322,35 @@ tensorflow::Status InsertNodesInDb(const int64 num_artifact_nodes,
   return tensorflow::Status::OK();
 }
 
+int64 GenerateRandomNumberFromUD(const UniformDistribution& dist,
+                                 std::minstd_rand0& gen) {
+  std::uniform_int_distribution<int64> uniform_dist{dist.minimum(),
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: update the function,  see some doc about the random distribution

https://abseil.io/docs/cpp/guides/random#reuse-generators-when-possible

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 6]
```diff
@@ -265,4 +322,35 @@ tensorflow::Status InsertNodesInDb(const int64 num_artifact_nodes,
...
+                                                    dist.maximum()};
+  return uniform_dist(gen);
+}
+
+std::discrete_distribution<int64>
+GenerateCategoricalDistributionWithDirichletPrior(
+    const int64 sample_size, const int64 concentration_param) {
+  std::default_random_engine generator;
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: pass a real generator & reuse the generation

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 7]
```diff
@@ -265,4 +322,35 @@ tensorflow::Status InsertNodesInDb(const int64 num_artifact_nodes,
...
+  std::gamma_distribution<double> gamma_distribution(concentration_param, 1.0);
+  std::vector<double> weights(sample_size);
+  for (int64 i = 0; i < sample_size; ++i) {
+    weights[i] = gamma_distribution(generator);
+  }
+  // Renormalization.
+  double sum = std::accumulate(weights.begin(), weights.end(), 0.0);
+  for (int64 i = 0; i < sample_size; ++i) {
+    weights[i] /= sum;
+  }
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: drop this?

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 8]
```diff
@@ -265,4 +322,35 @@ tensorflow::Status InsertNodesInDb(const int64 num_artifact_nodes,
...
+  // Uses these random number generated w.r.t. a Dirichlet distribution with
+  // `concentration_param` to represent the possibility of being chosen for each
+  // integer within [0, sample_size) in a discrete distribution.
+  std::discrete_distribution<int64> discrete_distribution{weights.begin(),
+                                                          weights.end()};
+  return discrete_distribution;
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: return constructor

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 9]
```diff
@@ -132,38 +177,50 @@ tensorflow::Status GetExistingTypes(const FillNodesConfig& fill_nodes_config,
...
+
+tensorflow::Status GetExistingNodes(
+    const FillContextEdgesConfig& fill_context_edges_config,
+    MetadataStore& store, std::vector<Node>& existing_non_context_nodes,
+    std::vector<Node>& existing_context_nodes) {
+  FetchNode fetch_non_context_node;
+  FetchNode fetch_context_node = FetchContext;
+  switch (fill_context_edges_config.specification()) {
+    case FillContextEdgesConfig::ATTRIBUTION: {
+      fetch_non_context_node = FetchArtifact;
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: drop this line? ditto for other places

File: ml_metadata/tools/mlmd_bench/util.h [hunk 10]
```diff
@@ -65,6 +76,17 @@ tensorflow::Status InsertNodesInDb(int64 num_artifact_nodes,
...
+// Generates random integer within the range of the specified `dist`.
+int64 GenerateRandomNumberFromUD(const UniformDistribution& dist,
+                                 std::minstd_rand0& gen);
+
+// Generates a categorical distribution with a Dirichlet prior with
+// `concentration_param`, the sample size of the returned discrete distribution
+// is specified by `sample_size`.
+std::discrete_distribution<int64>
+GenerateCategoricalDistributionWithDirichletPrior(int64 sample_size,
+                                                  int64 concentration_param);
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.h]: move to workload .cc, or add test coverage

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 11]
```diff
@@ -72,14 +72,6 @@ tensorflow::Status GetAndValidateExistingTypes(
   return tensorflow::Status::OK();
 }
 
-// Generates random integer within the range of specified `dist`.
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: revert the change on this file

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+  const double non_context_alpha =
+      fill_context_edges_config_.non_context_node_popularity()
+          .dirichlet_alpha();
+  const double context_alpha =
+      fill_context_edges_config_.context_node_popularity().dirichlet_alpha();
+  // Generates categorical distribution with a Dirichlet distribution with
+  // a concentrate parameter.
+  std::discrete_distribution<int64> non_context_node_index_dist =
+      GenerateCategoricalDistributionWithDirichletPrior(
+          existing_non_context_nodes.size(), non_context_alpha);
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc]: pass a config instead of alpha

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+  std::discrete_distribution<int64> context_node_index_dist =
+      GenerateCategoricalDistributionWithDirichletPrior(
+          existing_context_nodes.size(), context_alpha);
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc]: ditto

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+
+  std::minstd_rand0 gen(absl::ToUnixMillis(absl::Now()));
+
+  for (int64 i = 0; i < num_operations_; ++i) {
+    curr_bytes = 0;
+    PutAttributionsAndAssociationsRequest put_request;
+    const int64 num_edges =
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc]: move the distribution outside of the for

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+    std::discrete_distribution<int64>& non_context_node_index_dist,
+    std::discrete_distribution<int64>& context_node_index_dist,
+    MetadataStore& store,
+    std::unordered_map<int64, std::unordered_set<int64>>& unique_checker,
+    PutAttributionsAndAssociationsRequest& request, int64& curr_bytes) {
+  CHECK((std::is_same<T, Attribution>::value ||
+         std::is_same<T, Association>::value))
+      << "Unexpected Types";
+  int64 i = 0;
+  std::default_random_engine gen;
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc]: pass in the generator

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+  while (i < num_edges) {
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc:131]: + comment on top fo the while, metioning rejection sampling

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+    const int64 non_context_node_index = non_context_node_index_dist(gen);
+    const int64 context_node_index = context_node_index_dist(gen);
+    int64 non_context_node_id;
+    if (std::is_same<T, Attribution>::value) {
+      non_context_node_id =
+          absl::get<Artifact>(
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc]: create a function passing the Attribution/Association as template

none_context_node_id = get_non_context_node(Attribution () ..?

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+        absl::get<Context>(existing_context_nodes[context_node_index]).id();
+
+    bool already_existed_in_current_setup =
+        CheckDuplicateContextEdgeInCurrentSetUp(
+            non_context_node_id, context_node_id, unique_checker);
+    tensorflow::Status status = CheckDuplicateContextEdgeInDb<T>(
+        non_context_node_id, context_node_id, store);
+    if (!status.ok() && status.code() != tensorflow::error::ALREADY_EXISTS) {
+      return status;
+    }
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc]: drop case

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+    // Rejection sampling.
+    if (already_existed_in_current_setup || !status.ok()) {
+      continue;
+    }
+    SetCurrentContextEdge<T>(non_context_node_id, context_node_id, request);
+    // Increases `curr_bytes` with the size of `non_context_node_id` and 
+    // `context_node_id`.
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc]: write why this is 8*2

**Review by hughmiao — CHANGES_REQUESTED** [review]
thanks, Brian.

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.h (hunk 2 above)
Comment (BrianSong) [reply ml_metadata/tools/mlmd_bench/fill_context_edges_workload.h]: Will use a absl::flat_hash_map<int64, std::vector<int64>> here since std::pair cannot be automatically hashed.

File: ml_metadata/tools/mlmd_bench/util.cc (hunk 5 above)
Comment (BrianSong) [reply ml_metadata/tools/mlmd_bench/util.cc]: Extracted this function. One line function is indeed not necessary. Thanks :)

File: ml_metadata/tools/mlmd_bench/util.h (hunk 10 above)
Comment (BrianSong) [reply ml_metadata/tools/mlmd_bench/util.h]: Moved to workload.cc. Plan to add test coverage in the next PR. Thanks!

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc (hunk 12 above)
Comment (BrianSong) [reply ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc]: Yes, since GenerateCategoricalDistributionWithDirichletPrior() will also be used in FillEvents workload later, I think it may be better to pass in a alpha instead of a concrete type of config. Wdyt?

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc (hunk 13 above)
Comment (BrianSong) [reply ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc]: Discussed above.

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc (hunk 18 above)
Comment (BrianSong) [reply ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc]: Yes, for this one, it maybe necessary since in the tests, I also included a case where the db has pre-inserted context edges. So, we should also perform a checking inside db to pass the tests. Wdyt?

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc (hunk 17 above)
Comment (BrianSong) [reply ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc]: Good point!

File: ml_metadata/tools/mlmd_bench/benchmark_test.cc [hunk 20]
```diff
@@ -80,6 +80,24 @@ TEST(BenchmarkTest, CreatWorkloadTest) {
...
+            }
+            workload_configs: {
+              fill_context_edges_config: {
+                specification: ASSOCIATION
+                non_context_node_popularity: { dirichlet_alpha: 1 }
+                context_node_popularity: { dirichlet_alpha: 1 }
+                num_edges: { minimum: 1 maximum: 10 }
+              }
+              num_operations: 150
+            }
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/benchmark_test.cc:139]: split the test, w. util

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+// `non_context_node_id` and `context_node_id`.
+template <typename T>
+void SetCurrentContextEdge(const int64 non_context_node_id,
+                           const int64 context_node_id,
+                           PutAttributionsAndAssociationsRequest& request) {
+  if (std::is_same<T, Attribution>::value) {
+    Attribution* context_edge = request.add_attributions();
+    context_edge->set_artifact_id(non_context_node_id);
+    context_edge->set_context_id(context_node_id);
+  } else if (std::is_same<T, Association>::value) {
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_context_edges_workload.cc]: nit, else

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload_test.cc [hunk 22]
```diff
@@ -0,0 +1,295 @@
...
+namespace ml_metadata {
+namespace {
+
+constexpr int kNumberOfOperations = 100;
+constexpr int kNumberOfExistedTypesInDb = 132;
+constexpr int kNumberOfExistedNodesInDb = 145;
+constexpr int kNumberOfExistedContextEdgesInDb = 105;
+constexpr int kNumberOfEdgesPerRequest = 5;
+
+constexpr auto config_str = R"(
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_context_edges_workload_test.cc]: drop auto, use types

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload_test.cc [hunk 23]
```diff
@@ -0,0 +1,295 @@
...
+      default:
+        LOG(FATAL) << "Wrong specification for FillContextEdges!";
+    }
+  }
+  return tensorflow::Status::OK();
+}
+
+// Inserts `num_attributions` attributions and `num_associations` associations
+// inside db. Returns detailed error if query executions failed.
+tensorflow::Status InsertContextEdgesInDb(const int64 num_attributions,
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_context_edges_workload_test.cc]: drop?

**Review by hughmiao — CHANGES_REQUESTED** [review]
thanks, Brian. some comments before approval.

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload_test.cc (hunk 22 above)
Comment (BrianSong) [reply ml_metadata/tools/mlmd_bench/fill_context_edges_workload_test.cc]: Got you. As for the `auto` in `fill_nodes_workload_test.cc` and `fill_types_workload_test.cc`, will change them accordingly along with the later TODOs in order to decrease the number of changed files in this PR.

File: ml_metadata/tools/mlmd_bench/fill_context_edges_workload_test.cc (hunk 23 above)
Comment (BrianSong) [reply ml_metadata/tools/mlmd_bench/fill_context_edges_workload_test.cc]: Still keep this function but set up and run another FillContextEdges workload inside as we discussed. Thanks! :)

**Review by hughmiao — APPROVED** [review]
thanks, Brian
